### ISSUE TITLE
Cannot install "error[E0658]: use of unstable library feature 'thread_local_state': state querying was recently added (see issue #27716)"

### ISSUE THREAD

**Issue by bortzmeyer** [jedisct1/rust-doh#16]
```
% cargo install doh-proxy
...
   Compiling tokio-threadpool v0.1.7
   Compiling parking_lot v0.6.4
   Compiling tokio-reactor v0.1.6
error[E0658]: use of unstable library feature 'thread_local_state': state querying was recently added (see issue #27716)
   --> /home/stephane/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.6/src/sharded_rwlock.rs:158:18
    |
158 |     REGISTRATION.try_with(|reg| reg.index).unwrap_or(0)
    |                  ^^^^^^^^

error: aborting due to previous error

error: Could not compile `tokio-reactor`.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `doh-proxy v0.1.10`, intermediate artifacts can be found at `/tmp/cargo-install.rmjXAghhsfKL`

Caused by:
  build failed
```

Complete n00b in Rust so I'm afraid I cannot find the cause myself.

```
%  cargo --version
cargo 0.26.0
% rustc --version
rustc 1.25.0
```

**Comment by jedisct1** [comment]
You probably need a more recent version of Rust.

The current stable version is 1.29.1.

**Comment by bortzmeyer** [comment]
@jedisct1 You mean I need a "rolling release" OS like Arch Linux and not Debian stable or Ubuntu LTS :-(

**Comment by jedisct1** [comment]
You can use [rustup](https://rustup.rs), which is probably how most people are installing Rust.

Rust is still a moving target, especially when it comes to async I/O, so even common libraries don't support old versions any more

**Comment by Glandos** [comment]
In Debian, [rustc](https://packages.debian.org/sid/rustc) is available in version 1.29.0 in unstable, 1.28.0 in testing, and 1.24.1 in stable. This is quite remarkable for a fast moving target like rust.

And, in my personal opinion, you should stop using rustup, which is yet another random shell script downloaded from the Internet, running as root on your local machine, and installing binaries and libraries in some places that you may not remember.
Don't get me wrong, rustup is maybe well written, it's just that we cannot review everything before trusting it.

**Comment by dariusc93** [comment]
Though this issue a coupe of months old, I do want to comment that rustup doesnt run as root. Everything is installed in your home directory within `.cargo`. Compiler, toolchain, etc., all are there.
